### PR TITLE
(wip) Refactor EWMH window activation

### PR DESCRIPTION
**Pull request by liskin**
### Description

Work in progress on #396.

### Checklist

  - [X] I've read [CONTRIBUTING.md](https://github.com/xmonad/xmonad/blob/master/CONTRIBUTING.md)

  - [ ] I tested my changes with [xmonad-testing](https://github.com/xmonad/xmonad-testing)

  - [ ] I updated the `CHANGES.md` file

  - [ ] I updated the `XMonad.Doc.Extending` file (if appropriate)

### CODE REVIEW

**Comment by liskin** [comment]
(test failures expected, we need to merge https://github.com/xmonad/X11/pull/71 and release X11)

**Comment by TheMC47** [comment]
I played around with this a bit, and it's definitely an improvement to the UX in my opinion. Awesome work :tada: I'm not familiar with `X.H.Focus` or `X.A.WorkspaceNames` so I didn't get to try them fully (my config is still pretty basic), but I have some suggestions for the documentation of `XMonad.Hooks.EwmhDesktops` module:

- The examples in your personal dotfiles are nice, especially the one with `activateHook`.  I'm only a bit confused on why one would need `workspaceListSort`
- Regarding fullscreen support:
  -  It might be nice to mention why you *wouldn't* want fullscreen support. I personally don't use it, because I want picture-in-picture support sometimes. So it's maybe a good idea to mention that and refer to `windowedFullscreenFixEventHook`
  -  Since we're making the window float, if the user focuses another window in the same workspace by mistake, they'll have no indication that they're focused on the wrong window. Maybe it makes sense to have a combinator that we can use on `windows W.focusUp` for example, so it only triggers if no window in the current workspace is in fullscreen mode. It doesn't have to be here, I'm just thinking out loud

**Comment by liskin** [comment]
> * The examples in your personal dotfiles are nice, especially the one with `activateHook`.  I'm only a bit confused on why one would need `workspaceListSort`

Yeah, I'll definitely add examples. Basically the current state of the PR is that I haven't even tried to change the docs. :-(

Good point about `workspaceListSort` — I don't actually know why it's there. I think one use might be to hide scratchpad workspaces, but there may be other use cases that would be nice to mention, but I have absolutely no idea about any of these. May need to ask around…

> * Regarding fullscreen support:
>   
>   * It might be nice to mention why you _wouldn't_ want fullscreen support. I personally don't use it, because I want picture-in-picture support sometimes. So it's maybe a good idea to mention that and refer to `windowedFullscreenFixEventHook`

Will do.

>   * Since we're making the window float, if the user focuses another window in the same workspace by mistake, they'll have no indication that they're focused on the wrong window. Maybe it makes sense to have a combinator that we can use on `windows W.focusUp` for example, so it only triggers if no window in the current workspace is in fullscreen mode. It doesn't have to be here, I'm just thinking out loud

There's https://hackage.haskell.org/package/xmonad-contrib-0.16/docs/XMonad-Layout-Fullscreen.html#v:fullscreenFocus which probably deals with this problem. I've never used X.L.Fullscreen personally, but it's an interesting approach if one needs more control than just floats. Oh, and X.L.Fullscreen is another usecase of disabling fullscreen support in ewmh, as the event hooks would conflict otherwise.

Anyway, I intend to add (un)fullscreening hooks to EwmhConfig to make X.L.Fullscreen integrate better, to improve unfullscreening of floated windows, and possibly other fullscreen improvements: I wrote more about this in https://github.com/xmonad/xmonad-contrib/pull/395#issuecomment-773232202
But I get the feeling that fullscreening hooks should probably be in a separate PR, and I should try to reduce the scope of this one (#399): refactor the activation and X.H.Focus stuff, finish docs, review, merge, and then deal with the rest.

**Comment by TheMC47** [comment]
> I should try to reduce the scope of this one (#399): refactor the activation and X.H.Focus stuff, finish docs, review, merge, and then deal with the rest.

that sounds reasonable :+1: tell me if you need help with/feedback on the docs (not sure whether I'll have useful things to say about the implementation)

**Comment by liskin** [comment]
I cherry-picked the non-controversial bits into master over the last few days. I intend to reimplement the rest on top of the extensible config stuff (#547) later.

**Comment by sgf-dma** [comment]
In fact, i don't really understand what review you want from me. I have nothing to say about code, because i already forget (again) how all this work. Or you just want, that i try my config on current master?

**Comment by liskin** [comment]
@sgf-dma Once this is in a usable state (right now it's a bit behind current developments), I'd welcome if you tried using it in your config. I remember you had some concerns about this being unwieldy for certain use cases. If you don't remember or don't care, though, I'm fine with that, and I'll just try my best not to break legitimate use cases.

tl;dr: I'll let you know once it's ready.

**Comment by sgf-dma** [comment]
@liskin Ok, thanks. Well, use cases, which i thought essential (when wrote X.H.Focus) are provided in examples in comments. So, it'd be good, if all examples will work.

**Comment by liskin** [comment]
(just a rebase so that I can my daily driver, nothing exciting yet)
